File: timer.js
/**
 * Focus timer for Produco
 * JavaScript implementation of the focus timer
 */

class FocusTimer {
  constructor(workDuration = 25, breakDuration = 5) {
    this.workDuration = workDuration * 60; // Convert to seconds
    this.breakDuration = breakDuration * 60;
    this.timeRemaining = this.workDuration;
    this.isRunning = false;
    this.isPaused = false;
    this.interval = null;
    this.onTick = null;
    this.onComplete = null;
    this.currentTask = null;
  }

  start(task = null) {
    if (this.isRunning) return false;
    
    this.isRunning = true;
    this.isPaused = false;
    this.currentTask = task;
    this.timeRemaining = this.workDuration;
    
    this.interval = setInterval(() => {
      this.tick();
    }, 1000);
    
    return true;
  }

  pause() {
    if (!this.isRunning || this.isPaused) return false;
    
    this.isPaused = true;
    clearInterval(this.interval);
    return true;
  }

  resume() {
    if (!this.isRunning || !this.isPaused) return false;
    
    this.isPaused = false;
    this.interval = setInterval(() => {
      this.tick();
    }, 1000);
    return true;
  }

  stop() {
    if (!this.isRunning) return null;
    
    clearInterval(this.interval);
    const session = {
      task: this.currentTask,
      duration: (this.workDuration - this.timeRemaining) / 60,
      completed: this.timeRemaining <= 0
    };
    
    this.isRunning = false;
    this.isPaused = false;
    this.timeRemaining = this.workDuration;
    this.currentTask = null;
    
    return session;
  }

  tick() {
    this.timeRemaining--;
    
    if (this.onTick) {
      this.onTick(this.timeRemaining);
    }
    
    if (this.timeRemaining <= 0) {
      this.complete();
    }
  }

  complete() {
    clearInterval(this.interval);
    this.isRunning = false;
    
    if (this.onComplete) {
      this.onComplete();
    }
  }

  getRemainingTime() {
    return this.timeRemaining;
  }

  getFormattedTime() {
    const minutes = Math.floor(this.timeRemaining / 60);
    const seconds = this.timeRemaining % 60;
    return `${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
  }

  isWorkSession() {
    return this.timeRemaining > 0;
  }
}

module.exports = FocusTimer;

File: timer.cpp
/**
 * Focus timer for Produco - C++ implementation
 * High-precision timing for focus sessions
 */

#include <iostream>
#include <chrono>
#include <thread>
#include <atomic>
#include <functional>

class FocusTimer {
private:
    std::atomic<bool> isRunning;
    std::atomic<bool> isPaused;
    int workDuration;
    int breakDuration;
    int timeRemaining;
    std::string currentTask;
    std::function<void()> onTick;
    std::function<void()> onComplete;
    std::thread timerThread;
    
public:
    FocusTimer(int workMin = 25, int breakMin = 5) 
        : isRunning(false), isPaused(false), 
          workDuration(workMin * 60), breakDuration(breakMin * 60),
          timeRemaining(workDuration) {
        std::cout << "Focus timer initialized" << std::endl;
    }
    
    ~FocusTimer() {
        stop();
    }
    
    bool start(const std::string& task = "") {
        if (isRunning) return false;
        
        isRunning = true;
        isPaused = false;
        currentTask = task;
        timeRemaining = workDuration;
        
        timerThread = std::thread([this]() {
            while (isRunning && timeRemaining > 0) {
                if (!isPaused) {
                    std::this_thread::sleep_for(std::chrono::seconds(1));
                    timeRemaining--;
                    
                    if (onTick) {
                        onTick();
                    }
                } else {
                    std::this_thread::sleep_for(std::chrono::milliseconds(100));
                }
            }
            
            if (timeRemaining <= 0 && onComplete) {
                onComplete();
            }
        });
        
        return true;
    }
    
    bool pause() {
        if (!isRunning || isPaused) return false;
        isPaused = true;
        return true;
    }
    
    bool resume() {
        if (!isRunning || !isPaused) return false;
        isPaused = false;
        return true;
    }
    
    void stop() {
        isRunning = false;
        if (timerThread.joinable()) {
            timerThread.join();
        }
    }
    
    int getTimeRemaining() const {
        return timeRemaining;
    }
    
    std::string getFormattedTime() const {
        int minutes = timeRemaining / 60;
        int seconds = timeRemaining % 60;
        char buffer[10];
        snprintf(buffer, sizeof(buffer), "%02d:%02d", minutes, seconds);
        return std::string(buffer);
    }
    
    bool getIsRunning() const {
        return isRunning;
    }
    
    void setOnTick(std::function<void()> callback) {
        onTick = callback;
    }
    
    void setOnComplete(std::function<void()> callback) {
        onComplete = callback;
    }
};

int main() {
    FocusTimer timer;
    
    timer.setOnTick([&]() {
        std::cout << "Time remaining: " << timer.getFormattedTime() << std::endl;
    });
    
    timer.setOnComplete([&]() {
        std::cout << "Focus session complete!" << std::endl;
    });
    
    timer.start("Test task");
    
    // Run for 5 seconds then stop
    std::this_thread::sleep_for(std::chrono::seconds(5));
    timer.stop();
    
    return 0;
}
